core/shared/src/main/scala/zio/prelude/Semiring.scala

```diff
@@ -0,0 +1,398 @@
+    zipWith(that)((a, _) => a)
+
+  /**
+   * Combines this collection of events with that collection of events to
+   * return the Cartesian product of events, keeping only the events from that
+   * collection.
+   */
+  final def zipRight[Z1 >: Z <: Unit, B](that: Semiring[Z1, B]): Semiring[Z1, B] =
+    zipWith(that)((_, b) => b)
+
+  /**
+   * Combines this collection of events with that collection of events to
+   * return the Cartesian product of events using the specified function.
+   */
+  final def zipWith[Z1 >: Z <: Unit, B, C](that: Semiring[Z1, B])(f: (A, B) => C): Semiring[Z1, C] =
+    self.flatMap(a => that.map(b => f(a, b)))
+}
+
+object Semiring {
+
+  final case class Both[+Z <: Unit, +A](left: Semiring[Z, A], right: Semiring[Z, A]) extends Semiring[Z, A] { self =>
+    override def equals(that: Any): Boolean                                         =
+      that match {
+        case that: Semiring[Unit, Any] =>
+          equal(that) || symmetric(associate)(self, that) || symmetric(distribute)(self, that) || commute(
+            that
+          ) || symmetric(empty)(self, that)
+        case _                         =>
+          false
+      }
+    override def hashCode: Int                                                      =
+      Semiring.flatten(self) match {
+        case Nil                         => Empty.hashCode
+        case set :: Nil if set.size == 1 => set.head.hashCode
+        case seq                         => seq.hashCode
+      }
+    private def associate(l: Semiring[Unit, Any], r: Semiring[Unit, Any]): Boolean  =
+      (l, r) match {
+        case (Both(Both(al, bl), cl), Both(ar, Both(br, cr))) =>
+          al == ar && bl == br && cl == cr
+        case _                                                =>
+          false
+      }
+    private def commute(that: Semiring[Unit, Any]): Boolean                         =
+      (self, that) match {
+        case (Both(al, bl), Both(ar, br)) =>
+          al == br && bl == ar
+        case _                            =>
+          false
+      }
+    private def distribute(l: Semiring[Unit, Any], r: Semiring[Unit, Any]): Boolean =
+      (l, r) match {
+        case (Both(Then(al1, bl), Then(al2, cl)), Then(ar, Both(br, cr))) =>
+          al1 == al2 && al1 == ar && bl == br && cl == cr
+        case (Both(Then(al, cl1), Then(bl, cl2)), Then(Both(ar, br), cr)) =>
+          cl1 == cl2 && al == ar && bl == br && cl1 == cr
+        case _                                                            =>
+          false
+      }
+    private def equal(that: Semiring[Unit, Any]): Boolean                           =
+      (self, that) match {
+        case (br: Both[Unit, Any], bl: Both[Unit, Any]) =>
+          br.left == bl.left && br.right == bl.right
+        case _                                          =>
+          false
+      }
+  }
+
+  case object Empty extends Semiring[Unit, Nothing]
```

Comment:
Pattern matches against `Empty` don't work well because it is not known to be a valid subtype if `Z` is abstract, since `Z` could be `Nothing`. Maybe if we go down this route we should make the constructors private and only expose them through folding and the operators we provide?